common/minimodel/__init__.py

```diff
@@ -1008,3 +1013,52 @@ def total_used(self) -> timedelta:
 
     def observations(self) -> List[Observation]:
         return self.root_group.observations()
+
+
+class Plans:
+    """
+    A collection of Nightly Plan from a corresponding Site
+    """
+    def __init__(self, night_events):
+        self.durations = [len(night_events.times[idx]) for idx, _ in enumerate(night_events.time_grid)]
+        self.start_times = [night_events.local_times[idx][0] for idx, _ in enumerate(night_events.time_grid)]
+        self.end_times = [night_events.local_times[idx][-1] for idx, _ in enumerate(night_events.time_grid)]
+        self.time_slot_length = night_events.time_slot_length
+
+    def __iter__(self):
+        self.plans = []
+        return self
+
+    def __next__(self):
+        if len(self.plans) <= len(self.durations):
+            self.plans.append(Plan(next(iter(self.start_times)),
+                                   next(iter(self.end_times)),
+                                   slot_length=self.time_slot_length,
+                                   night_duration=next(iter(self.durations))))
+            return next(iter(self.plans))
+        raise StopIteration
+
+
+class Plan:
+    """
+    A 'plan' is a collection of nighly plans
+    """
+    def __init__(self, start, end, slot_length=1, night_duration=10):
+        self.start = start
+        self.end = end
+        self._time_slot_length = slot_length
+        self._time_slots_left = night_duration
+        self._visits = []
```

Comment:
This could easily be made a `dataclass` to be much simpler.
Also, why is the time slot length private? Won't this be needed?

```
@dataclass
class Plan:
    start: datetime
    end: datetime
    _time_slot_length: timedelta = timedelta(minutes=1)
    _time_slots_left: int = 10  # I'm not sure why 10 is chosen here?

    def __post_init__(self):
        self._visits = []
```

Reply:
Oh, one more concern: if a `Plan` is a collection of nightly plans, the number of time slots for each night will be different. So how will `_time_slots_left` work?